### PR TITLE
TypeScript rewrite

### PR DESCRIPTION
**Pull request by MidSpike**
~~# Draft, do not merge yet~~

Tested, works

### CODE REVIEW

File: src/bot/commands/verify.ts [hunk 1]
```diff
@@ -47,8 +50,8 @@ module.exports = {
                     new Discord.MessageEmbed({
                         color: 0xFFFF00,
                         author: {
-                            iconURL: `${client.user.displayAvatarURL({ dynamic: true })}`,
-                            name: `${client.user.username} | Verification System`,
+                            iconURL: `${client.user!.displayAvatarURL({ dynamic: true })}`,
+                            name: `${client.user!.username} | Verification System`,
                         },
                         title: 'You are already verified!',
```
Comment (TotoTheEgg) [inline src/bot/commands/verify.ts:56]: Change to `you can not verify, you are already verified` or something along the lines of "you cant verify" so users stop spamming the verify command after they have already verified

File: src/bot/handlers/logs/guild_member_retention.ts [hunk 2]
```diff
@@ -2,32 +2,26 @@
...
     const member_retention_logging_channel = await client.channels.fetch(member_retention_logging_channel_id);
     if (!member_retention_logging_channel) throw new Error('Failed to fetch logging channel');
+    if (!member_retention_logging_channel.isText()) throw new TypeError('member_retention_logging_channel is not a text channel');
 
     const user_creation_timestamp = getMarkdownFriendlyTimestamp(member.user.createdTimestamp);
-    const member_joined_timestamp = getMarkdownFriendlyTimestamp(member.joinedTimestamp);
+    const member_joined_timestamp = getMarkdownFriendlyTimestamp(member.joinedTimestamp ?? Date.now());
 
     await member_retention_logging_channel.send({
         embeds: [
```
Comment (TotoTheEgg) [inline src/bot/handlers/logs/guild_member_retention.ts:27]: Change ID to user mention, and remove account join date to clean up and reduce unnecessary information.

File: src/bot/handlers/logs/guild_member_retention.ts [hunk 3]
```diff
@@ -52,17 +46,15 @@ async function guildMemberAddLogger(member) {
...
     const member_retention_logging_channel = await client.channels.fetch(member_retention_logging_channel_id);
     if (!member_retention_logging_channel) throw new Error('Failed to fetch logging channel');
+    if (!member_retention_logging_channel.isText()) throw new TypeError('member_retention_logging_channel is not a text channel');
 
     const user_creation_timestamp = getMarkdownFriendlyTimestamp(member.user.createdTimestamp);
-    const member_joined_timestamp = getMarkdownFriendlyTimestamp(member.joinedTimestamp);
+    const member_joined_timestamp = getMarkdownFriendlyTimestamp(member.joinedTimestamp ?? Date.now());
     const member_left_timestamp = getMarkdownFriendlyTimestamp(Date.now());
 
     await member_retention_logging_channel.send({
```
Comment (TotoTheEgg) [inline src/bot/handlers/logs/guild_member_retention.ts:60]: Remove account leave date to reduce cludder

File: src/bot/commands/downloads.ts [hunk 4]
```diff
@@ -2,23 +2,24 @@
...
     permission_level: command_permission_levels.PUBLIC,
     cooldown: 10_000,
-    async execute(message, args) {
+    async execute(
+        message: Discord.Message<true>,
+        args: {
+            [key: string]: unknown;
+        },
+    ) {
         message.channel.send({
```
Comment (TotoTheEgg) [inline src/bot/commands/downloads.ts:23]: Not sure how plausible this is, but reply to someone's message if it contains more than the download/retrieve/get command informing them they don't need to run the command for every product they own?

File: src/bot/commands/lookup.ts [hunk 5]
```diff
@@ -66,7 +68,7 @@ module.exports = {
                     new Discord.MessageEmbed({
```
Comment (TotoTheEgg) [inline src/bot/commands/lookup.ts:68]: add "display user profile" to the embed if the user exists in the database, that info is more helpful while on mobile to be able to copy user IDs from the mention, as well as get the roblox profile link

**Review by TotoTheEgg — COMMENTED** [review]
Not sure where to put this comment, but an issue I have seen with the cooldown is when a user runs a command on cooldown, even if there are 2 seconds left on the cooldown it will reset the cooldown, is there a way to have the command cooldown not reset every time the command is run?

File: src/bot/handlers/logs/guild_member_retention.ts (hunk 2 above)
Comment (MidSpike) [reply src/bot/handlers/logs/guild_member_retention.ts:27]: won't change, intended features

File: src/bot/commands/downloads.ts (hunk 4 above)
Comment (MidSpike) [reply src/bot/commands/downloads.ts:23]: won't fix

File: src/bot/commands/lookup.ts (hunk 5 above)
Comment (MidSpike) [reply src/bot/commands/lookup.ts:68]: won't fix, use the profile command

**Comment by MidSpike** [comment]
@TotoTheEgg please keep unrelated concerns out of this pull request, this pull request is focused on rewriting the language used by our bot to TypeScript.

**Comment by TotoTheEgg** [comment]
> @TotoTheEgg please keep unrelated concerns out of this pull request, this pull request is focused on rewriting the bot in TypeScript.

Gotcha, I will submit a new review in the morning.

**Review by WillFJS — APPROVED** [review]
All looks good to me

**Comment by TotoTheEgg** [comment]
Looks swagger to me 👍

**Review by rossloika — APPROVED** [review]
After spending about 40 minutes reviewing, and running 'tsc --watch' with the current commit. I have found no issues. Approved! 👍

**Review by TotoTheEgg — APPROVED** [review]
Swaggie swagger updates my g

**Pull request merged**